### PR TITLE
Replace FRC20 with SRC20

### PR DESCRIPTION
**Pull request by DefiCake**

### CODE REVIEW

File: packages/fungible-token/bridge-fungible-token/Forc.toml [hunk 1]
```diff
@@ -7,3 +7,4 @@ name = "bridge_fungible_token"
 [dependencies]
 contract_message_receiver = { path = "../../message-predicates/contract-message-receiver" }
 reentrancy = { git = "https://github.com/FuelLabs/sway-libs", tag = "v0.12.0" }
+src_20 = { git = "https://github.com/FuelLabs/sway-standards", tag = "v0.1.2" }
```
Comment (Braqzen) [inline packages/fungible-token/bridge-fungible-token/Forc.toml:10]:
```suggestion
src_20 = { git = "https://github.com/FuelLabs/sway-standards", tag = "v0.1.2" }

```

File: packages/fungible-token/bridge-fungible-token/src/interface/bridge.sw [hunk 2]
```diff
@@ -1,5 +1,7 @@
 library;
 
+use std::string::String;
```
Comment (K1-R1) [inline packages/fungible-token/bridge-fungible-token/src/interface/bridge.sw:3]: I dont think this is used in the file
```suggestion
```

File: packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        interface::src20::{decimals, name, symbol, total_supply},
+        setup::{
+            create_msg_data, create_wallet, get_asset_id, relay_message_to_contract,
+            setup_environment, BridgeFungibleTokenContract, BridgingConfig,
+        },
+    };
+    use fuels::accounts::wallet::WalletUnlocked;
+
+    /// This setup mints tokens so that they are registered as minted assets in the bridge
+    async fn setup_test() -> BridgeFungibleTokenContract<WalletUnlocked> {
```
Comment (K1-R1) [inline packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs:15]: Can you move this to utils?

File: packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            utxo_inputs.message[0].clone(),
+            utxo_inputs.contract,
+            &utxo_inputs.coin[..],
+        )
+        .await;
+
+        contract
+    }
+
+    #[ignore]
```
Comment (K1-R1) [inline packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs:53]: This test is ignored? 
It looks like the comment about the SDK is outdated and so this test can be un-ignored and the assertion added back in?

File: packages/fungible-token/bridge-fungible-token/src/interface/bridge.sw [hunk 5]
```diff
@@ -1,5 +1,7 @@
...
+
```
Comment (Braqzen) [inline packages/fungible-token/bridge-fungible-token/src/interface/bridge.sw:4]:
```suggestion
```

Why was this added?

File: packages/fungible-token/bridge-fungible-token/src/main.sw [hunk 6]
```diff
@@ -197,27 +198,32 @@ impl Bridge for Contract {
     }
 }
 
-impl FRC20 for Contract {
+impl SRC20 for Contract {
     #[storage(read)]
-    fn total_supply() -> U256 {
-        let minted: u64 = storage.tokens_minted.try_read().unwrap_or(0);
+    fn total_assets() -> u64 {
```
Comment (Braqzen) [inline packages/fungible-token/bridge-fungible-token/src/main.sw:203]: Why are the assets and total supply 0 / None?

File: packages/fungible-token/bridge-fungible-token/src/main.sw [hunk 7]
```diff
@@ -257,3 +263,8 @@ fn _asset_to_sub_id(asset_id: AssetId) -> b256 {
     require(sub_id.is_some(), BridgeFungibleTokenError::AssetNotFound);
     sub_id.unwrap()
 }
+
+#[storage(read)]
+fn _check_asset_is_minted_by_contract(asset_id: AssetId) {
+    require(storage.asset_to_sub_id.get(asset_id).try_read().is_some(), BridgeFungibleTokenError::AssetNotFound);
+}
```
Comment (Braqzen) [inline packages/fungible-token/bridge-fungible-token/src/main.sw:270]: I would remove this because it's only used twice and it has a single require. If it did a bit more then I would potentially keep it

File: packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    #[tokio::test]
+    async fn check_total_supply() {
+        // Lacking SDK support on version 0.43
```
Comment (Braqzen) [inline packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs:56]: SDK is way ahead of this version

File: packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            &utxo_inputs.coin[..],
+        )
+        .await;
+
+        contract
+    }
+
+    #[ignore]
+    #[tokio::test]
+    async fn check_total_supply() {
```
Comment (Braqzen) [inline packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs:55]: total_assets test?

File: packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs (hunk 3 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs]: Done

File: packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs (hunk 4 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs]: Fixed

File: packages/fungible-token/bridge-fungible-token/src/main.sw (hunk 6 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/src/main.sw:205]: FIxed

File: packages/fungible-token/bridge-fungible-token/src/main.sw (hunk 7 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/src/main.sw]: Removed

File: packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs (hunk 8 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs]: Outdated, resolved

File: packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs (hunk 9 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/tests/functions/src20/mod.rs]: Done

File: packages/fungible-token/bridge-fungible-token/src/main.sw [hunk 10]
```diff
@@ -201,28 +203,45 @@ impl Bridge for Contract {
...
-        None
+    fn total_supply(asset: AssetId) -> Option<u64> {
+        if storage.asset_to_sub_id.get(asset).try_read().is_none() {
+            return None;
+        }
+
+        // Reasoning: an extreme supply token such as SHIBA could overflow the
+        // total supply, which is defined by SRC20 as u64.
+        // .as_u64() will revert on overflow, in that case, it will return u64::max() 
+        Some(storage.tokens_minted.get(asset).read().as_u64().unwrap_or(u64::max()))
```
Comment (Braqzen) [inline packages/fungible-token/bridge-fungible-token/src/main.sw:218]: So this will technically be an incorrect value under an overflow. Unsure about this

File: packages/fungible-token/bridge-fungible-token/tests/functions/bridge/mod.rs [hunk 11]
```diff
@@ -16,6 +16,7 @@ mod success {
 
     use super::*;
 
+    use crate::utils::interface::src20::total_supply;
```
Comment (Braqzen) [inline packages/fungible-token/bridge-fungible-token/tests/functions/bridge/mod.rs:19]: I would always merge imports.

What happens if you run `cargo clippy --all-features --all-targets -- -D warnings`?

**Review by Braqzen — COMMENTED** [review]
@bitzoic what do you think about the SRC20 here? It's a little different

File: packages/fungible-token/bridge-fungible-token/tests/functions/bridge/mod.rs (hunk 11 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/tests/functions/bridge/mod.rs:19]: Complains about other things, not this one specifically

```
   Compiling fuel-contract-message-predicate v0.3.0 (/Users/mad/fuel/bridge/packages/message-predicates/contract-message-predicate)
    Checking bridge-fungible-token-contract v0.3.0 (/Users/mad/fuel/bridge/packages/fungible-token/bridge-fungible-token)
error: `assert!(false, ..)` should probably be replaced
   --> packages/fungible-token/bridge-fungible-token/tests/functions/bridge/mod.rs:347:17
    |
347 |                 assert!(false, "Transaction did not succeed")
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: use `panic!(..)` or `unreachable!(..)`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#assertions_on_constants
    = note: `-D clippy::assertions-on-constants` implied by `-D warnings`

error: `assert!(false, ..)` should probably be replaced
    --> packages/fungible-token/bridge-fungible-token/tests/functions/message_receiver/mod.rs:1185:17
     |
1185 |                 assert!(false, "Transaction did not revert");
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = help: use `panic!(..)` or `unreachable!(..)`
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#assertions_on_constants
```

My go to command with clippy is `cargo clippy --all-features --all-targets --fix`, I 'll make sure to run this one as well.

I 'll push a commit to merge this one, and fix the others that -D warnings gives too

Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/tests/functions/bridge/mod.rs]: Fixed

File: packages/fungible-token/bridge-fungible-token/src/main.sw (hunk 10 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/src/main.sw]: Fixed: it now registers a refund when the total_supply cap is reached

File: packages/fungible-token/bridge-fungible-token/src/main.sw [hunk 12]
```diff
@@ -52,7 +57,8 @@ configurable {
 storage {
     asset_to_sub_id: StorageMap<AssetId, b256> = StorageMap {},
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/src/main.sw:58]: I see this is only used in the withdrawal function of the bridge, wouldn't it make more sense to pass it as an argument to the `withdraw(to: b256, sub_id: SubId)` function and assert that the `msg_asset_id() == AssetId::new(contract_id(), sub_id)`?

File: packages/fungible-token/bridge-fungible-token/src/main.sw [hunk 13]
```diff
@@ -197,27 +214,46 @@ impl Bridge for Contract {
...
     #[storage(read)]
-    fn name() -> str[64] {
-        NAME
+    fn name(asset: AssetId) -> Option<String> {
+        if storage.asset_to_sub_id.get(asset).try_read().is_none() {
+            return None;
+        }
+
+        Some(String::from_ascii_str(from_str_array(NAME)))
     }
```
Comment (bitzoic) [inline packages/fungible-token/bridge-fungible-token/src/main.sw:239]:
```suggestion
    fn name(asset: AssetId) -> Option<String> {
        match storage.total_supply.get(asset).try_read() {
            Some(v) => Some(String::from_ascii_str(from_str_array(NAME))),
            None => None,
        }        
    }
```

**Review by bitzoic — COMMENTED** [review]
With the SRC-20 standard you no longer need the empty spaces at the end of your configurable for `NAME` and `SYMBOL`. If every bridged token is using the same name then just "MyToken" or "BridgedAsset" will suffice. I'm also unsure if every asset bridged uses the same decimals i.e. NFTs and high supply assets but I don't know the specific implementation that is happening here.

File: packages/fungible-token/bridge-fungible-token/src/main.sw [hunk 14]
```diff
@@ -88,6 +92,20 @@ impl MessageReceiver for Contract {
...
+
+                let current_total_supply = storage.tokens_minted.get(asset_id).try_read().unwrap_or(0);
+                let new_total_supply = current_total_supply + amount;
+
+                if new_total_supply < current_total_supply {
+                    register_refund(message_data.from, message_data.token_address, message_data.token_id, message_data.amount);
+                    return;
+                }
+
+                enable_panic_on_overflow();
```
Comment (Braqzen) [inline packages/fungible-token/bridge-fungible-token/src/main.sw:105]: Wondering what happens if we return above but do not reach this line. Do we continue to have disabled panics for overflow or are they re-enabled automatically in the subsequent calls?

File: packages/fungible-token/bridge-fungible-token/src/main.sw (hunk 12 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/src/main.sw:58]: This would imply that callers have to be aware of the `sub_id` of the asset they are withdrawing. There has been a lot of discussion around this and I think we have not yet reached an agreement, as you said, it might be that the `sub_id` might be available as part of the message in the future. For the time being, and given that block space remains cheap, the consensus is that we keep the mapping of asset_id <-> sub_id in storage so that callers do not need to be aware of the sub_id.

I do not have a strong opinion on this though.

File: packages/fungible-token/bridge-fungible-token/src/main.sw (hunk 14 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/src/main.sw:105]: To the best of my knowledge, this function cannot be called from other contexts inside the contract (it acts as an external function in solidity). 

OTOH, I 'd assume that contract-to-contract calls cannot "poison" these kind of flag configurations.

File: packages/fungible-token/bridge-fungible-token/src/main.sw (hunk 12 above)
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/src/main.sw]: I 've also refactored the use of `b256` types and replaced it with SubId and AssetId where needed

File: packages/fungible-token/bridge-fungible-token/tests/functions/message_receiver/mod.rs [hunk 15]
```diff
@@ -120,19 +121,244 @@ mod success {
...
+            amount,
+            token_address,
+            from,
+            token_id: _,
+        } = refund_registered_events[0];
+
+        // Check logs
+        assert_eq!(amount, Bits256(encode_hex(max_deposit_amount)));
+        assert_eq!(token_address, Bits256::from_hex_str(BRIDGED_TOKEN).unwrap());
+        assert_eq!(from, Bits256::from_hex_str(FROM).unwrap());
```
[the remark this answers is not in the document]
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/tests/functions/message_receiver/mod.rs:343]: Does not hurt to do so 👍🏻  I just maintained consistency across tests

File: packages/fungible-token/bridge-fungible-token/src/main.sw [hunk 16]
```diff
@@ -192,32 +209,46 @@ impl Bridge for Contract {
...
-        U256::from((0, 0, 0, minted))
+    #[storage(read)]
+    fn total_supply(asset: AssetId) -> Option<u64> {
+        storage.tokens_minted.get(asset).try_read()
     }
 
     #[storage(read)]
-    fn name() -> str[64] {
-        NAME
+    fn name(asset: AssetId) -> Option<String> {
```
[the remark this answers is not in the document]
Comment (DefiCake) [reply packages/fungible-token/bridge-fungible-token/src/main.sw:229]: For context, let's remember that an instance of this sway contract can only hold one EVM token, 
- with ERC20 tokens using a token_id = sub_id = 0, which translates as having an unique AssetId that can have any amount from 1 to U64.max
- with ERC721 using token_id for each collectible, which translates as having multiple AssetIds that can have only 1 unit each, but still all belong to the same "collection" that is ERC721, and all of these have the same symbol

Illustration:
- ERC20: you are bridging 10 USDT and that translates to a single AssetId say 0x0..abcd, with amount 10. Then, if you query the name for AssetId 0x0..abcd, it will return "Fuel USDT"
- ERC721: you are bridging Ape 1, Ape 2 and Ape 3 with token_id's 1, 2, 3. That translates to AssetId 0x0..1, 0x0..2, 0x0..3, with amount 1 for each one of them. Then, if you query any of these assets for its name, it should return the same name: "Fuel Ape", which is the original ERC721 collection.

**Review by Braqzen — APPROVED** [review]
